src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfTestCase.cs

```diff
@@ -40,11 +43,41 @@ internal WcfTestCase(IXunitTestCase testCase, string skippedReason)
 
         public void Deserialize(IXunitSerializationInfo info) { _testCase.Deserialize(info); }
 
-        public Task<RunSummary> RunAsync(
+        public async Task<RunSummary> RunAsync(
             IMessageSink diagnosticMessageSink, IMessageBus messageBus, object[] constructorArguments,
             ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource)
         {
-            return new XunitTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource).RunAsync();
+            ConcurrentQueue<EventWrittenEventArgs> events = new ConcurrentQueue<EventWrittenEventArgs>();
+	          s_testListener._eventWritten = events.Enqueue;
+
+            var runsummary = await (new XunitTestCaseRunner(this, DisplayName, _testCase.SkipReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource)).RunAsync();
+
+            if (runsummary.Failed > 0)
+            {
+                s_testListener._eventWritten = null;
+                Console.WriteLine(string.Format("---ETW Trace for Test {0} Begins---", DisplayName));
+                foreach (var item in events)
+                {
+                    try
+                    {
+                        Console.WriteLine(string.Format(item.Message, item.Payload));
+                    }
+                    catch (Exception)
+                    {
+                        Console.WriteLine(string.Format("ETW message: {0}, payload below was received", item.Message));
+                        foreach (var payloadPara in item.Payload)
```

Comment:
The payload isn't big from my testing result, we don't usually include very large message in the functional test. The risk is also mitigated by we only output ETW for failing cases.

Agree to collect them to a StringBuilder and do a single WriteLine.